drivers/gpio/gpio_lpc11u6x.c

```diff
@@ -0,0 +1,655 @@
+	struct lpc11u6x_gpio_regs *gpio_regs = (struct lpc11u6x_gpio_regs *)
+		(config->shared->gpio_base + LPC11U6X_GPIO_REGS);
+
+	gpio_regs->not[config->port_num] = pins;
+
+	return 0;
+}
+
+/**
+ * @brief Attach a free interrupt line to a GPIO.
+ *
+ * @param shared   Pointer to a structure shared between all the GPIO ports.
+ * @param intpin   GPIO port and pin numbers encoded into a value compatible
+ *                 with the INTPIN register (included in the PINTSEL register).
+ *
+ * @retval >0      Number of the attached interrupt on success.
+ * @retval -EBUSY  All the interrupt lines are already attached.
+ */
+static int
+pintsel_attach(const struct gpio_lpc11u6x_shared *shared, uint8_t intpin)
+{
+	uint8_t irq;
+	int ret = -EBUSY;
+	uint32_t *pintsel_reg =
+		(uint32_t *) (shared->syscon_base + LPC11U6X_PINTSEL_REGS);
+
+	for (irq = 0; irq < shared->nirqs; irq++) {
+		/* GPIO already attached. */
+		if ((pintsel_reg[irq] & BIT_MASK(5)) == intpin) {
+			return irq;
+		}
+
+		if (ret < 0 && (pintsel_reg[irq] & BIT_MASK(5)) == 0) {
+			ret = irq;
+		}
+	}
+	/* Attach GPIO to the first free interrupt found if any. */
+	if (ret >= 0) {
+		pintsel_reg[ret] = intpin;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Detach an interrupt line from a GPIO.
+ *
+ * @param shared   Pointer to a structure shared between all the GPIO ports.
+ * @param intpin   GPIO port and pin numbers encoded into a value compatible
+ *                 with the INTPIN register (included in the PINTSEL register).
+ *
+ * @retval >0      Number of the detached interrupt on success.
+ * @retval -EINVAL No attached interrupt found for the requested GPIO.
+ */
+static int
+pintsel_detach(const struct gpio_lpc11u6x_shared *shared, uint8_t intpin)
+{
+	uint8_t irq;
+	uint32_t *pintsel_reg =
+		(uint32_t *) (shared->syscon_base + LPC11U6X_PINTSEL_REGS);
+
+	for (irq = 0; irq < shared->nirqs; irq++) {
+		if ((pintsel_reg[irq] & BIT_MASK(5)) == intpin) {
+			pintsel_reg[irq] = 0;
+			return irq;
+		}
+	}
+	return -EINVAL;
+}
+
+static int gpio_lpc11u6x_pin_interrupt_configure(struct device *port,
+						 gpio_pin_t pin,
+						 enum gpio_int_mode mode,
+						 enum gpio_int_trig trig)
+{
+	const struct gpio_lpc11u6x_config *config = DEV_CFG(port);
+	struct gpio_lpc11u6x_data *data = DEV_DATA(port);
+	struct lpc11u6x_pint_regs *pint_regs = (struct lpc11u6x_pint_regs *)
+		(config->shared->gpio_base + LPC11U6X_PINT_REGS);
+	uint8_t intpin;
+	int irq;
+
+	if (pin >= config->ngpios) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Because the PINTSEL register only have 6 bits to encode a pin
+	 * number, then PIO2_8 to PIO2_23 can't be attacehd to an interrupt
+	 * line.
+	 */
+	if (config->port_num == 2 && pin > 7) {
+		return -ENOTSUP;
+	}
+
+	/*
+	 * Convert the requested GPIO port and pin numbers into a value
+	 * compatible with the INTPIN register (included in the PINTSEL
+	 * register).
+	 */
+	intpin = pin;
+	if (config->port_num == 1) {
+		intpin += 24;
```

Comment:
This looks similar to https://github.com/zephyrproject-rtos/zephyr/blob/4b194eb4fc40723bba46acf0fae36d5d1d1e59a8/drivers/gpio/gpio_mcux_lpc.c#L199 

Could this be reused?

Reply:
Hi Andrei,

Many thanks for spending some time on this PR.

In the gpio_mcux_lpc driver each free interrupt can be attached _one_ time to a pin. It is never detached. In all the dtsi files using the "nxp,lpc-gpio" binding, two GPIOs ports are defined and 8 interrupts lines are split equally between this two port. Basically each GPIO port gets four interrupts attached statically at built time. Once the 4 interrupts lines are consumed/used, it is done.

On the LPC11U6x SoC there are 8 interrupts lines available as well, but 3 GPIOs ports. This doesn't divide equally. In addition each of the 8 interrupts can be used by each of the GPIO ports. So we choose a dynamic approach. At run time any of the port can attach any of the free interrupts lines (as longer there is one available). An when an interrupt is disabled then it is detached from the pin. So the interrupt line becomes available again for any of the GPIO ports. It is done by the pintsel_detach function. In the gpio_mcux_lpc driver and in the NXP MCUx HAL, there is no support to detach interrupts from pin. To support the attach/detach feature the pintsel_attach function have to run some checks on the pintsel registers. attach_pin_to_isr() does not. That's why I think that attach_pin_to_isr should not be reused here.

Reply:
Thank you Simon for your answer. You are right, pins are not changed at runtime, once attached they are not detached, this comes from allocating resources statically in device tree. If you need to detach the pin at runtime it wouldn't be a good ideea to switch to the other driver.
The allocation of 4 interrupt lines/port is arbitrary and should not be taken as hardwired, most of the code in gpio_mcux_lpc is designed so interrupts can be differently distributed depending on application. At the time I implemented that I had LPC55s69 an LPC54114 in mind, the allocation is fair and with no particular application in mind.